### PR TITLE
add HTTPURI for more performant uri access

### PR DESCRIPTION
**Pull request by tanner0101**
Motivation:

Currently the only way to access a request head's uri is via a computed property that converts it to a String. The internal storage however is usually a byte buffer. Being able to access the uri bytes without going through String would be a performance win.

Modifications:

Removed URI in favor or HTTPURI and internal HTTPURIStorage

Result:

HTTP URI can be accessed more performantly.

See https://github.com/apple/swift-nio/issues/342

### CODE REVIEW

**Comment by weissi** [comment]
@swift-nio-bot test this please

File: Sources/NIOHTTP1/HTTPTypes.swift [hunk 1]
```diff
@@ -115,21 +124,53 @@ public struct HTTPRequestHead: Equatable {
...
-/// Internal representation of a URI
-enum URI {
-    case string(String)
-    case byteBuffer(ByteBuffer)
-}
+/// Represents a the URI component of an `HTTPRequestHead`. Allows the URI
+/// to be accessed as a `String` or as raw bytes.
+public struct HTTPURI: ExpressibleByStringLiteral {
+    /// Internal storage.
+    var storage: HTTPURIStorage
```
Comment (Lukasa) [inline Sources/NIOHTTP1/HTTPTypes.swift]: This can probably be private.

File: Sources/NIOHTTP1/HTTPTypes.swift [hunk 2]
```diff
@@ -115,21 +124,53 @@ public struct HTTPRequestHead: Equatable {
...
+            }
+        }
+        set {
+            storage = .string(newValue)
+        }
+    }
+}
+
+/// Internal representation of a URI
+enum HTTPURIStorage {
```
Comment (Lukasa) [inline Sources/NIOHTTP1/HTTPTypes.swift]: This guy can probably be an inner type on `HTTPURI`, which also means we can avoid the implicit name spacing in the type name.

**Comment by tanner0101** [comment]
@Lukasa thanks for the comments, updated!

File: Sources/NIOHTTP1/HTTPTypes.swift [hunk 3]
```diff
@@ -115,19 +124,51 @@ public struct HTTPRequestHead: Equatable {
...
+    public var string: String {
+        get {
+            switch storage {
+            case .string(let s): return s
+            case .byteBuffer(let b): return b.getString(at: b.readerIndex, length: b.readableBytes)!
+            }
+        }
+        set {
+            storage = .string(newValue)
         }
```
[the remark this answers is not in the document]
Comment (weissi) [reply Sources/NIOHTTP1/HTTPTypes.swift]: @Lukasa this one is fine as it's defined on something that doesn't do CoW. It were a problem if storage would be a `class` and we'd find a `if !isKnownUniquelyReferenced(&self.storage)` in here

File: Sources/NIOHTTP1/HTTPTypes.swift [hunk 4]
```diff
@@ -115,19 +124,51 @@ public struct HTTPRequestHead: Equatable {
...
-/// Internal representation of a URI
-enum URI {
-    case string(String)
-    case byteBuffer(ByteBuffer)
-}
+/// Represents the URI component of an `HTTPRequestHead`. Allows the URI
+/// to be accessed as a `String` or as raw bytes.
+public struct HTTPURI: ExpressibleByStringLiteral {
+    /// Internal representation of a URI
+    enum _Storage {
```
[the remark this answers is not in the document]
Comment (weissi) [reply Sources/NIOHTTP1/HTTPTypes.swift]: this is a total nit which we can ignore: today we use `_Storage` only (?) for things that are classes inside of structs to implement CoW and that has certain benefits as you can easily search for it in Xcode. I'd prefer to have this named literally anything else :)

File: Sources/NIOHTTP1/HTTPTypes.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (Lukasa) [reply Sources/NIOHTTP1/HTTPTypes.swift]: 👍 Got it thanks.

**Comment by tanner0101** [comment]
@weissi for the majority of routes in Vapor, you never need to access the URI as a `String`. The router prefers to work with the URI as a collection of bytes and that is the only place where the URI is actually accessed. I don't know how expensive converting the URI to a String actually _is_ but I think its non-trivial. Also Vapor is just going to need to convert that `String` back to bytes which as you mentioned is another allocation.

**Comment by weissi** [comment]
@tanner0101 making a plain ASCII string from a number of bytes should be quite cheap (CC @gottesmm) besides the allocation.
Regarding the Vapor routing, why don't you work on `string.utf8` that is also a collection of bytes and should also be really cheaply accessed on ascii only strings as the String should internally store `[UInt8]` for ASCII only strings as far as I understand.

**Comment by helje5** [comment]
@weissi:
> My assumption is that in any application at some point the URI will need to become a String anyway

Even ignoring the charset issue, the full URL or even just a URL path is rarely needed as a String. Even if you just consider the path of a request, you essentially have to split and decode its components (for special chars and normalization):

    GET /Projects%2FImportant/My%20Document.gif

This is much better kept as a ByteBuffer and transformed to what the higher level actually wants later, often - but not necessarily - a:

    let path : String = [ "/", "Projects/Important" / "My Document.gif" ]
    
It kinda is a waste to alloc the URIs as Strings when we could just pass through a ByteBuffer slice, which is then later decoded by the web framework into the structure it desires (enums for the matching routes seems to be popular too, that would potentially involve 0 allocs going from network to the app level).

File: Sources/NIOHTTP1/HTTPTypes.swift [hunk 5]
```diff
@@ -115,19 +124,51 @@ public struct HTTPRequestHead: Equatable {
...
+        self.storage = storage
+    }
+
+    /// See `ExpressibleByStringLiteral`.
+    public init(stringLiteral value: String) {
+        storage = .string(value)
+    }
+
+    /// Access this `HTTPURI`'s raw bytes.
+    /// - Parameter closure: A closure accepting an `UnsafeRawBufferPointer` to this URI's raw bytes.
```
[the remark this answers is not in the document]
Comment (helje5) [reply Sources/NIOHTTP1/HTTPTypes.swift]: To be honest I still don't understand why you think that this should ever be a String. I would back the HTTPURI by just a ByteBuffer(/Slice). Probably add a `stringValue` property which creates a String on demand for logging and such. Maybe also add a `pathComponents` property (or an iterator) which does the splitting and percent decoding.

Rational: string of a full path never needed except for logging, octet-preservation, zero copy for proxies, one malloc less for the common case.

**Comment by weissi** [comment]
@tanner0101 a benchmark would be great and I agree a `String` API should be there. However @helje5 's use-case is a valid one too so even though I was against it at the beginning I now think we will need to offer a `Collection` of `UInt8` API too.  And as I said before I think we should let the user choose if they want us to eagerly decode the URI to a `String` or not. If you write a proxy that needs to accept all sorts of rubbish you don't want the URI to be decoded.
In Vapor's use case it depends if String is fast enough or not I guess.

**Comment by weissi** [comment]
that's not a thing anymore, `String` is not fast and we parse the URL eagerly :)

**Comment by milseman** [comment]
> that's not a thing anymore, String is not fast and we parse the URL eagerly :)

"Not fast" or "now fast"? If the former, please file a bug with a micro benchmark and I'll make it faster.

**Comment by normanmaurer** [comment]
@milseman I am pretty sure its `now fast` ;)

**Comment by weissi** [comment]
@milseman so sorry, what a typo. Definitely _now_ fast :).

**Pull request closed without merging**